src/libraries/System.Private.CoreLib/src/System/Globalization/InvariantModeCasing.cs

```diff
@@ -0,0 +1,779 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using System.Runtime.CompilerServices;
+using Internal.Runtime.CompilerServices;
+
+namespace System.Globalization
+{
+    internal static partial class InvariantModeCasing
+    {
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static char ToLower(char c)
+        {
+           Debug.Assert(GlobalizationMode.Invariant);
+
+           ushort v = LowerCase1[c >> 8];
+           v = LowerCase2[v + ((c >> 4) & 0xF)];
+           v = LowerCase3[v + (c & 0xF)];
+
+           return v == 0 ? c : (char)v;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static char ToUpper(char c)
+        {
+           Debug.Assert(GlobalizationMode.Invariant);
+
+           ushort v = UpperCase1[c >> 8];
+           v = UpperCase2[v + ((c >> 4) & 0xF)];
+           v = UpperCase3[v + (c & 0xF)];
+
+           return v == 0 ? c : (char)v;
+        }
+
+        internal static unsafe string ToLower(string s)
+        {
+            if (s.Length == 0)
+            {
+                return string.Empty;
+            }
+
+            fixed (char* pSource = s)
```

Comment:
Are pointers necessary for performance here?  Wondering if we could use span and if most of the bounds checks could be eliminated by the JIT.

Reply:
I'll try that.